### PR TITLE
Draft: common: fix sycc420_to_rgb buffer overflow

### LINKED ISSUE #1347: Heap-buffer-overflow in color.c:379:42 in sycc420_to_rgb
Hi,
I found a vulnerability in current master [0bda718](https://github.com/uclouvain/openjpeg/commit/0bda7188b7b545232a341f1d978b1e4feda46fc2), and I also reproduced it on latest released version [v2.4.0](https://github.com/uclouvain/openjpeg/releases/tag/v2.4.0).

**Crash Summary**
A heap-buffer-overflow in color.c:379:42 in sycc420_to_rgb, it can lead to heap-based buffer overflow via a crafted `.j2k` file when decompress it.

**Crash Analysis**
There is insufficient validation of `*cb`.
https://github.com/uclouvain/openjpeg/blob/0bda7188b7b545232a341f1d978b1e4feda46fc2/src/bin/common/color.c#L375-L381

**PoC:**
[poc.j2k.gz](https://github.com/uclouvain/openjpeg/files/6402272/poc.j2k.gz)

**To reproduce (x86-64 Ubuntu 20.04.2 with gcc 9.3.0):**
```bash
CFLAGS='-g -fsanitize=address' cmake .. -DCMAKE_BUILD_TYPE=Release
make

./bin/opj_decompress -i ./poc.j2k -o out.png
```

**ASAN report:**
```
=================================================================
==2371124==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x612000000760 at pc 0x0000004f278c bp 0x7ffd11a3eca0 sp 0x7ffd11a3ec98
READ of size 4 at 0x612000000760 thread T0
    #0 0x4f278b in sycc420_to_rgb /home/yuawn/fuzz-targets/openjpeg/reproduce/openjpeg/src/bin/common/color.c:379:42
    #1 0x4f278b in color_sycc_to_rgb /home/yuawn/fuzz-targets/openjpeg/reproduce/openjpeg/src/bin/common/color.c:416:9
    #2 0x4cb136 in main /home/yuawn/fuzz-targets/openjpeg/reproduce/openjpeg/src/bin/jp2/opj_decompress.c:1589:13
    #3 0x7f653bef10b2 in __libc_start_main /build/glibc-YbNSs7/glibc-2.31/csu/../csu/libc-start.c:308:16
    #4 0x41d4fd in _start (/home/yuawn/fuzz-targets/openjpeg/reproduce/openjpeg/build/bin/opj_decompress+0x41d4fd)

0x612000000760 is located 0 bytes to the right of 288-byte region [0x612000000640,0x612000000760)
allocated by thread T0 here:
    #0 0x498027 in posix_memalign (/home/yuawn/fuzz-targets/openjpeg/reproduce/openjpeg/build/bin/opj_decompress+0x498027)
    #1 0x7f653c38aa5f in opj_aligned_alloc_n /home/yuawn/fuzz-targets/openjpeg/reproduce/openjpeg/src/lib/openjp2/opj_malloc.c:61:9
    #2 0x7f653c38aa5f in opj_aligned_malloc /home/yuawn/fuzz-targets/openjpeg/reproduce/openjpeg/src/lib/openjp2/opj_malloc.c:209:12
    #3 0x7f653c37c257 in opj_alloc_tile_component_data /home/yuawn/fuzz-targets/openjpeg/reproduce/openjpeg/src/lib/openjp2/tcd.c:697:39
    #4 0x7f653c37c257 in opj_tcd_decode_tile /home/yuawn/fuzz-targets/openjpeg/reproduce/openjpeg/src/lib/openjp2/tcd.c:1561:18
    #5 0x7f653c2d6131 in opj_j2k_decode_tile /home/yuawn/fuzz-targets/openjpeg/reproduce/openjpeg/src/lib/openjp2/j2k.c:9727:11
    #6 0x7f653c2f275e in opj_j2k_decode_tiles /home/yuawn/fuzz-targets/openjpeg/reproduce/openjpeg/src/lib/openjp2/j2k.c:11568:15
    #7 0x7f653c2dba43 in opj_j2k_exec /home/yuawn/fuzz-targets/openjpeg/reproduce/openjpeg/src/lib/openjp2/j2k.c:8871:33
    #8 0x7f653c2dba43 in opj_j2k_decode /home/yuawn/fuzz-targets/openjpeg/reproduce/openjpeg/src/lib/openjp2/j2k.c:11871:11

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/yuawn/fuzz-targets/openjpeg/reproduce/openjpeg/src/bin/common/color.c:379:42 in sycc420_to_rgb
Shadow bytes around the buggy address:
  0x0c247fff8090: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
  0x0c247fff80a0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c247fff80b0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c247fff80c0: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
  0x0c247fff80d0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0c247fff80e0: 00 00 00 00 00 00 00 00 00 00 00 00[fa]fa fa fa
  0x0c247fff80f0: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
  0x0c247fff8100: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c247fff8110: 00 00 00 00 00 00 00 00 00 00 00 00 fa fa fa fa
  0x0c247fff8120: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
  0x0c247fff8130: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==2371124==ABORTING
```

### PR DESCRIPTION
**Pull request by msabwat**
This adds a check so that we do not increment Y/Cb/Cr components if we reach the last pixel.

Fixes #1347 

I am wondering if I should prevent the loop from reaching this point instead of adding a check?

Also it needs more testing, it does not overflow anymore, but I am not sure I did not break anything else.

### CODE REVIEW

**Comment by StayPirate** [comment]
Please consider https://github.com/uclouvain/openjpeg/issues/1347#issuecomment-858573954 before merging. Thanks.

**Comment by msabwat** [comment]
I did some tests, and they do not seem to add failed tests after the patch with or without asan 

img

@StayPirate done

**Comment by ValZapod** [comment]
There is also a problem that sYCC decoding is not bit perfect with ffmpeg libopenjpeg wrapper  (or native decoder jpeg2000 for that matter).

File: src/bin/common/color.c [hunk 1]
```diff
@@ -368,12 +368,18 @@ static void sycc420_to_rgb(opj_image_t *img)
...
-            ++y;
+            if (*y != img->comps[0].data[max]) {
+                ++y;
+            }
             ++r;
             ++g;
             ++b;
-            ++cb;
-            ++cr;
+            if (*cb != img->comps[1].data[max]) {
```
[the remark this answers is not in the document]
Comment (msabwat) [reply src/bin/common/color.c]: Thank you for the review, I push forced the small changes.

**Comment by rouault** [comment]
hum, so this breaks the NR-DEC-issue411-ycc420.jp2-115-decode-md5 test case of the test suite. Either my suggestion is wrong, either this test case needs refreshing

**Comment by msabwat** [comment]
> hum, so this breaks the NR-DEC-issue411-ycc420.jp2-115-decode-md5 test case of the test suite. Either my suggestion is wrong, either this test case needs refreshing

How can this work if we only increment one component of the pixel? are there cases where `sycc420_to_rgb` will just convert luma without bllue and red (it seems odd to me, but I may be wrong)

**Comment by ValZapod** [comment]
> will just convert luma without bllue and red (it

There is no blue and red. Also all grayscale is ONLY Y component.

**Comment by rouault** [comment]
> How can this work if we only increment one component of the pixel? are there cases where `sycc420_to_rgb` will just convert luma without bllue and red (it seems odd to me, but I may be wrong)

you only increment blue and red one out of 4 pixels due to the 4:2:0 subsampling.

**Comment by ValZapod** [comment]
> either this test case needs refreshing

You need to check whether those pixels that are wrong are out-of-gamut for BT.601 full range matrix (and that is actually a lot of triplets, obviously 255, y, x, where x and y are not 128 are one of those examples, but actually even in the middle a lot values are like that).

**Comment by rouault** [comment]
upon reflection, my suggestion to remove ++cb and ++cr completely seems wrong to me. The current code of master should be fine. I'm not sure why the overflow comes from, but the remedy should likely be something else

**Comment by msabwat** [comment]
> upon reflection, my suggestion to remove ++cb and ++cr completely seems wrong to me. The current code of master should be fine. I'm not sure why the overflow comes from, but the remedy should likely be something else

Sorry I didn't answer before, I wanted to test other things before I get back to you. Ran the test and the decoded image is :
out

There seems to be an additional line in the bottom ( I don't know if it's relevant), in the tests there is a comparison with the md5 hash, is it okay if I take the decoded file for the current master as a base ?

If I understand correctly, checking for the bounds is not enough, I need to investigate why the loop goes too far?

**Comment by StayPirate** [comment]
is there any update on this?

**Comment by msabwat** [comment]
I have a better understanding of subsampling, but I still need some time to understand the sample that causes the overflow and the issue in the code to fix it the right way.
I'll get back asap

**Comment by ZaquL** [comment]
> hum, so this breaks the NR-DEC-issue411-ycc420.jp2-115-decode-md5 test case of the test suite. Either my suggestion is wrong, either this test case needs refreshing

https://github.com/uclouvain/openjpeg-data/blob/master/input/nonregression/issue411-ycc420.jp2

This file decodes in -c:v jpeg2000 with 1 byte difference with -c:v libopenjpeg. One byte 0x57 is 0x56 in ffmpeg's native decoder.

[ZAQU.zip](https://github.com/uclouvain/openjpeg/files/8700166/ZAQU.zip)

**Comment by opoplawski** [comment]
Any progress here?

**Comment by opoplawski** [comment]
Any progress here?

**Comment by SimonWoidig** [comment]
Would appreciate if any update about this vulnerability was made!

**Comment by msabwat** [comment]
Sure, I will have a look and let you know

**Comment by msabwat** [comment]
After comparing with  ffmpeg's decoder, it seems the problem is about the EOC marker, which is missing. I am still reading openjpeg code to see if I can prevent the file from being decoded. Because if decoding ends, the image is processed with wrong data, which is causing the overflow. The initial fix which was to try to act on that is wrong.

I will test with another decoder, and will maybe have a new proposition of a fix for review by the end of this week.

**Comment by msabwat** [comment]
With Grok, it refuses to decompress the file for another reason:
```
[goose@blueduck build]$ ./bin/grk_decompress -i ../../openjpeg/poc.j2k -o test.jpg
[2023-11-24 17:28:42.724] [error] Non-compliant Rsiz value 0x2e00 in SIZ marker
[2023-11-24 17:28:42.724] [error] grk_decompress: failed to read the header
```

The file shouldn't be decompressed if:
- there is no EOC marker found
- Rsiz value is non compliant

@rouault, would it be acceptable to do this?

**Comment by rouault** [comment]
> would it be acceptable to do this?

I've the feeling the changes you suggest would just hide the root issue in sycc420_to_rgb() by rejecting the reproducer early. Unless they are really related to the issue in sycc420_to_rgb(), but a deeper analysis should be done to understand the root cause of the heap-buffer-overflow

**Comment by msabwat** [comment]
@rouault All right! will do. I went by ZaQul's earlier comments saying that ybcbr's code should not be touched, but if you feel it is wrong, I will get back to it. 

If the file is not valid, isn't it the role of the decoder to refuse the image if it does not comply with the spec?

**Comment by rouault** [comment]
> I went by ZaQul's earlier comments saying that ybcbr's code should not be touched, but if you feel it is wrong, I will get back to it.

There's no "sacred" code. If it is buggy, it should be fixed. I have no informed opinion where the source of the error is.

> If the file is not valid, isn't it the role of the decoder to refuse the image if it does not comply with the spec?

We already accept a number of non-compliances. This is on a case by case basis. Sometimes this can be just a small/pedantic non-compliance that can be recovered and helps decoding files that are available in the wild. Sometimes the file is just broken and we cannot do anything with it but rejecting it.
Here my point was that rejecting the reproducer file could still let the issue unfixed for other less broken files if the issue is in the sycc420_to_rgb code. But again we should have a detailed analysis of the root cause to see where the fix should be.

**Comment by msabwat** [comment]
> There's no "sacred" code.
I agree.

> But again we should have a detailed analysis of the root cause to see where the fix should be.
I will focus on that, thank you for taking the time to answer with such valuable information.

**Comment by deekal2** [comment]
Is this issue being actively looked into? This vulnerability is still unpatched after two years.

The last release for this component was 2 years ago. Is this component being maintained?

Is there an alternative to OpenJPEG?

**Comment by msabwat** [comment]
Closing since I don't have enough time to actively look into it.
